ZEngine/src/VulkanDevice.cpp

```diff
@@ -110,35 +92,63 @@ namespace ZEngine::Hardwares
         device_create_info.sType                = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
         device_create_info.queueCreateInfoCount = static_cast<uint32_t>(device_queue_create_info_collection.size());
         device_create_info.pQueueCreateInfos    = device_queue_create_info_collection.data();
-        device_create_info.enabledLayerCount    = 0;
 
         device_create_info.enabledExtensionCount   = static_cast<uint32_t>(device_extension_layer.size());
         device_create_info.ppEnabledExtensionNames = (device_extension_layer.size() > 0) ? device_extension_layer.data() : nullptr;
-        device_create_info.ppEnabledLayerNames     = nullptr;
+        auto layer                                 = std::array{"VK_LAYER_KHRONOS_validation"};
+        device_create_info.enabledLayerCount       = layer.size();
+        device_create_info.ppEnabledLayerNames     = layer.data();
         device_create_info.pEnabledFeatures        = nullptr;
         device_create_info.pNext                   = nullptr;
 
         ZENGINE_VALIDATE_ASSERT(vkCreateDevice(physical_device, &device_create_info, nullptr, &m_logical_device) == VK_SUCCESS, "Failed to create GPU logical device")
 
-        if (!m_graphic_queue_family_index_collection.empty())
-        {
-            vkGetDeviceQueue(m_logical_device, m_graphic_queue_family_index_collection[0], 0, &m_graphic_queue);
-        }
-
         if (!m_graphic_with_present_queue_family_index_collection.empty())
         {
-            vkGetDeviceQueue(m_logical_device, m_graphic_with_present_queue_family_index_collection[0], 0, &m_graphic_with_present_support_queue);
+            for (uint32_t family_index : m_graphic_with_present_queue_family_index_collection)
```

Comment:
You can do this but I would consider that overkill. On my machine, this code creates 16 queues for family index 0 alone, which I doubt you will ever need for your rendering workloads. I'm not sure what you intent to do but generally, I rarely see more than one queue created of a specific type (graphics, transfer, compute).

The typical workflow to parallelize Vulkan CPU work is rather to have multiple threads record command buffers in parallel and then submit them to a single queue.

I've also seen that you specify all queue families when creating resources and enable VK_SHARING_MODE_CONCURRENT so every queue you created can access them. Depening on the underlying Vulkan implementation and/or the hardware, this can result in lower performance than using a single queue and/or VK_SHARING_MODE_EXCLUSIVE.

I recommend you start with what you need for your current purposes, which is probably one graphics queue and maybe one transfer queue, if you want to offload staging and downloads from the graphics queue. But maybe just start with a graphics queue, you can do transfer there too.

Reply:
thanks a lot for the feedback.
I revisited the whole engine resource management to stick with only the needed queue (transfer, graphic) and maybe later compute